src/plugins/liveobjects/liveobject.ts

```diff
@@ -1,31 +1,80 @@
 import type BaseClient from 'common/lib/client/baseclient';
+import type EventEmitter from 'common/lib/util/eventemitter';
 import { LiveObjects } from './liveobjects';
 import { StateMessage, StateOperation } from './statemessage';
 import { DefaultTimeserial, Timeserial } from './timeserial';
 
+enum LiveObjectEvents {
+  Updated = 'Updated',
+}
+
 export interface LiveObjectData {
   data: any;
 }
 
-export abstract class LiveObject<T extends LiveObjectData = LiveObjectData> {
+export interface LiveObjectUpdate {
+  update: any;
+}
+
+export interface LiveObjectUpdateNoop {
+  // have optional update field with undefined type so it's not possible to create a noop object with a meaningful update property.
+  update?: undefined;
+  noop: true;
+}
+
+export interface SubscribeResponse {
+  unsubscribe(): void;
+}
+
+export abstract class LiveObject<
+  TData extends LiveObjectData = LiveObjectData,
+  TUpdate extends LiveObjectUpdate = LiveObjectUpdate,
+> {
   protected _client: BaseClient;
-  protected _dataRef: T;
+  protected _eventEmitter: EventEmitter;
+  protected _dataRef: TData;
   protected _objectId: string;
   protected _regionalTimeserial: Timeserial;
 
   constructor(
     protected _liveObjects: LiveObjects,
-    initialData?: T | null,
+    initialData?: TData | null,
     objectId?: string,
     regionalTimeserial?: Timeserial,
   ) {
     this._client = this._liveObjects.getClient();
+    this._eventEmitter = new this._client.EventEmitter(this._client.logger);
     this._dataRef = initialData ?? this._getZeroValueData();
     this._objectId = objectId ?? this._createObjectId();
     // use zero value timeserial by default, so any future operation can be applied for this object
     this._regionalTimeserial = regionalTimeserial ?? DefaultTimeserial.zeroValueTimeserial(this._client);
   }
 
+  subscribe(listener: (update: TUpdate) => void): SubscribeResponse {
+    this._eventEmitter.on(LiveObjectEvents.Updated, listener);
+
+    const unsubscribe = () => {
+      this._eventEmitter.off(LiveObjectEvents.Updated, listener);
+    };
+
+    return { unsubscribe };
```

Comment:
this is an interesting pattern, i'm not against it but i wonder if it's a good idea to add it here without adding it elsewhere for `channel.subscribe`, `presence.subscribe`, `EventEmitter.on`, etc?

Reply:
unfortunately it will be a breaking change for `channel.subscribe`, as it currently [returns the result](https://github.com/ably/ably-js/blob/18a255948c38d1e60715c8f5d6173369b57cb8d6/src/common/lib/client/realtimechannel.ts#L432) of the `channel.attach` call, a `ChannelStateChange` instance, if implicit attach is used (and it is by default).

we "could" add it to `presence.subsrcibe` and `EventEmitter.on` it seems, but we won't be able to keep parity in the API fully without this breaking change, and without specification changes.

Reply:
I do agree that it may be not the best option to have this API only in one place, and Mike raised the same concerns about API parity here https://ably.atlassian.net/wiki/spaces/LOB/pages/3354787841/LODR-019+Realtime+Client+SDK+public+API+-+Subscriptions?atl_f=PAGETREE#:~:text=For%20a%20Counter%2C%20we%20would%20get%20an%20update%20when%20it%27s%20been%20incremented., which is why your regular `liveObject.unsubscribe(listener)` call is still supported, just like `channel.unsubscribe(listener)` (however, I do not allow calling `liveObject.unsubscribe()` without arguments, as it is just straight up buggy and should not be the API for channels/presence neither)

at the same time it just feels wrong to not include a really nice developer API just to make sure we keep parity with other inferior API

Reply:
ok good point about the breaking change for `channel.subscribe`. i'd be interested to know why calling `unsubscribe()` without arguments is buggy but i suppose that doesn't affect this PR so will resolve the comment for now